src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs

```diff
@@ -86,55 +118,158 @@ public ValidationRuleSet(ValidationRuleSet ruleSet)
         /// <param name="rule">The rule.</param>
-        public void Add(ValidationRule rule)
+        /// <exception cref="OpenApiException">Exception thrown when rule already exists.</exception>
+        public void Add(string key, ValidationRule rule)
         {
-            if (!_rules.ContainsKey(rule.ElementType))
+            if (!_rules.ContainsKey(key))
             {
-                _rules[rule.ElementType] = new List<ValidationRule>();
+                _rules[key] = new List<ValidationRule>();
             }
 
-            if (_rules[rule.ElementType].Contains(rule))
+            if (_rules[key].Contains(rule))
             {
                 throw new OpenApiException(SRResource.Validation_RuleAddTwice);
             }
 
-            _rules[rule.ElementType].Add(rule);
+            _rules[key].Add(rule);
+        }
+
+        /// <summary>
+        /// Updates an existing rule with a new one.
+        /// </summary>
+        /// <param name="key">The key of the existing rule.</param>
+        /// <param name="newRule">The new rule.</param>
+        /// <param name="oldRule">The old rule.</param>
+        /// <returns>true, if the update was successful; otherwise false.</returns>
+        public bool Update(string key, ValidationRule newRule, ValidationRule oldRule)
+        {
+            if (!_rules.ContainsKey(key))
+            {
+                return false;
+            }
+            else
+            {
+                _rules[key].Remove(oldRule);
+                _rules[key].Add(newRule);
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Removes a collection of rules.
+        /// </summary>
+        /// <param name="key">The key of the collection of rules to be removed.</param>
+        /// <returns>true if the collection of rules with the provided key is removed; otherwise, false.</returns>
+        public bool Remove(string key)
+        {
+            return _rules.Remove(key);
+        }
+
+        /// <summary>
+        /// Removes a rule by key.
+        /// </summary>
+        /// <param name="key">The key of the rule to be removed.</param>
+        /// <param name="rule">The rule to be removed.</param>
+        /// <returns>true if the rule is successfully removed; otherwise, false.</returns>
+        public bool Remove(string key, ValidationRule rule)
+        {
+            if (_rules.TryGetValue(key, out IList<ValidationRule> validationRules))
+            {
+                return validationRules.Remove(rule);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Removes a rule from the list of rules.
+        /// </summary>
+        /// <param name="rule">The rule to be removed.</param>
+        /// <returns>true if the rule is successfully removed; otherwise, false.</returns>
+        public bool Remove(ValidationRule rule)
+        {
+            return _rules.Values.SelectMany(x => x).ToList().Remove(rule);
```

Comment:
```suggestion
            return _rules.Values.Remove(rule);
```
ToList means you're removing from a new list instance, effectively having no effect on the current data.